packages/envd/internal/api/store.go

```diff
@@ -2,23 +2,134 @@ package api
 
 import (
 	"encoding/json"
+	"fmt"
 	"net/http"
-
-	"github.com/rs/zerolog"
+	"slices"
 
 	"github.com/e2b-dev/infra/packages/envd/internal/host"
 	"github.com/e2b-dev/infra/packages/envd/internal/utils"
+	"github.com/e2b-dev/infra/packages/shared/pkg/keys"
+	"github.com/rs/zerolog"
+)
+
+const (
+	SigningReadOperation  = "read"
+	SigningWriteOperation = "write"
+)
+
+var (
+	accessTokenHeader = "X-Access-Token"
+
+	// paths that are always allowed without general authentication
+	allowedPaths = []string{
+		"GET/health",
+		"GET/files",
+		"POST/files",
+	}
 )
 
 type API struct {
-	logger  *zerolog.Logger
-	envVars *utils.Map[string, string]
+	logger      *zerolog.Logger
+	accessToken *string
+	envVars     *utils.Map[string, string]
+}
+
+type jsonErrorResponse struct {
+	Error string `json:"error"`
 }
 
 func New(l *zerolog.Logger, envVars *utils.Map[string, string]) *API {
 	return &API{logger: l, envVars: envVars}
 }
 
+func (a *API) WithAuthorization(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if a.accessToken != nil {
+			authHeader := req.Header.Get(accessTokenHeader)
+
+			// check if this path is allowed without authentication (e.g., health check, endpoints supporting signing)
+			allowedPath := slices.Contains(allowedPaths, req.Method+req.URL.Path)
+
+			if authHeader != *a.accessToken && !allowedPath {
+				response := jsonErrorResponse{Error: "Unauthorized access, please provide a valid access token or method signing if supported"}
+				responseBytes, _ := json.Marshal(response)
+
+				a.logger.Error().Msg("Trying to access secured envd without correct access token")
+
+				w.WriteHeader(http.StatusUnauthorized)
```

Comment:
How does this look like in SDK?

Reply:
For this feature to be used user will need to have new version of SDK with parameter supporting that so we can handle it properly i suppose?